clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -119,33 +116,259 @@ void BoundsAnalysis::CollectBoundsVars(const Expr *E, DeclSetTy &BoundsVars) {
+        return V;
   }
-  return false;
+
+  if (const auto *BO = dyn_cast<BinaryOperator>(E)) {
+    const auto *V = GetNtArrayVarDecl(BO->getLHS());
+    if (V)
+      return V;
+    return GetNtArrayVarDecl(BO->getRHS());
+  }
+  return nullptr;
+}
+
+void BoundsAnalysis::FillGenSet(const Expr *E, BoundsExpr *BE,
+                                ElevatedCFGBlock *EB,
+                                ElevatedCFGBlock *SuccEB) {
+
+  // TODO: Handle accesses of the form:
+  // "if (*(p + i) && *(p + j) && *(p + k))"
+
+  const auto *RBE = dyn_cast<RangeBoundsExpr>(BE);
+  if (!RBE)
+    return;
+
+  // The deref expr can be of 2 forms:
+  // 1. Ptr deref or array subscript: if (*p) or p[i]
+  // 2. BinaryOperator: if (*(p + e))
+
+  // SplitIntoBaseOffset tries to make the expr uniform. It returns a "Base"
+  // and an "Offset". Base is an expr containing only DeclRefExprs and Offset
+  // is an expr containing only IntegerLiterals. Here are some examples:
+  // Note: p and q are DelRefExprs, i and j are IntegerLiterals.
+
+  // Expr ==> Return Value
+  // p ==> (p, nullptr)
+  // p + i ==> (p, i)
+  // i + p ==> (p, i)
+  // p + i + j ==> (p, i + j)
+  // i + p + j + q ==> (p + q, i + j)
+
+  ExprPairTy DerefExprPair = SplitIntoBaseOffset(E);
+  const Expr *DerefBase = DerefExprPair.first;
+  const Expr *DerefOffset = DerefExprPair.second;
+  if (!DerefBase)
+    return;
+
+  // Also make the upper bounds expr uniform.
+  const Expr *UE = RBE->getUpperExpr();
+  if (!UE)
+    return;
+
+  ExprPairTy UpperExprPair = SplitIntoBaseOffset(UE);
+  const Expr *UpperBase = UpperExprPair.first;
+  const Expr *UpperOffset = UpperExprPair.second;
+  if (!UpperBase)
+    return;
+
+  // For bounds widening, the bases of the deref expr and the upper bounds expr
+  // should be the same. For example:
+  // _Nt_array_ptr<T> p : bounds(p, p + i);
+
+  // if (*(p + i)) // widen by 1
+  //   if (*(p + i + 1)) // widen by 2
+  //     if (*(p + i + 2)) // widen by 3
+
+  // if (*(p + i)) // widen by 1
+  //   if (*(p + i + 2)) // no widening
+
+  // if (*p) // no widening
+  //   if (*(p + 1)) // no widening
+  //     if (*(p + i)) // widen by 1
+  if (Lex.CompareExpr(DerefBase, UpperBase) !=
+      Lexicographic::Result::Equal)
+    return;
+
+  // Extract the ntptr from the lower bounds expr.
+  const VarDecl *V = GetNtArrayVarDecl(RBE->getLowerExpr());
+  if (!V)
+    return;
+
+  // Update the bounds of p on the edge EB->SuccEB only if we haven't already
+  // updated them.
+  if (EB->Gen[SuccEB->Block].count(V))
+    return;
+
+  const llvm::APInt Zero(Ctx.getTypeSize(Ctx.IntTy), 0);
+
+  const llvm::APInt DerefOffsetVal = !DerefOffset ? Zero :
+    dyn_cast<IntegerLiteral>(DerefOffset)->getValue();
+
+  const llvm::APInt UpperOffsetVal = !UpperOffset ? Zero :
+    dyn_cast<IntegerLiteral>(UpperOffset)->getValue();
+
+  // We cannot widen the bounds if the offset in the deref expr is less than
+  // the offset in the upper bounds expr. For example:
```

Comment:
Should we only widen the bounds only exactly when the offset for the deref expr equals the offset in the upper bounds expr?